### PR TITLE
Allow users to customize search algorithm in `DropdownMenu`

### LINKED ISSUE #136735: DropdownMenu shows incorrect currentHighlight
b/305662376

Customer reports the current highlight of their DropdownMenu is not correct.
I have not been able to reproduce this in an independent code sample, but the customer code and a video of the bug is available in the internal bug listed above.

Steps:

1. DropdownMenu has initial index of 0.
2. Customer opens menu, selects index 2, menu closes.
3. Customer opens menu, index 1 is highlighted.

@QuncCccccc do you have any thoughts on this? It looks like currentHighlight is incorrect. 

I think this could be the way we choose the currentHighlight based on filtering. The customer case has 3 items in the menu, in this order
* All
* Unread
* Read

The bug presents itself when `Read` is selected and the menu is reopened, and `Unread` is highlighted instead. At this point `Read` is in the text field of the menu, so we could be highlighting the first match to the String `read`. Maybe?

### PR DESCRIPTION
**Pull request by QuncCccccc**
Fixes #136735

This PR is to add a searchCallback to allow users to customize the search algorithm. This feature is used to fix b/305662376 which needs an exact match algorithm.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

**Comment by joshuauys** [comment]
Added bottom app bar to navigate between lists and profile page

File: packages/flutter/lib/src/material/dropdown_menu.dart [hunk 1]
```diff
@@ -303,6 +308,34 @@ class DropdownMenu<T> extends StatefulWidget {
   /// Defaults to null.
   final EdgeInsets? expandedInsets;
 
+  /// Invoked when [DropdownMenu.enableSearch] is true and return an index to
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/dropdown_menu.dart:311]:
```suggestion
  /// When  [DropdownMenu.enableSearch] is true this callback is used to compute the index of the search result to be highlighted.
```

File: packages/flutter/lib/src/material/dropdown_menu.dart [hunk 2]
```diff
@@ -303,6 +308,34 @@ class DropdownMenu<T> extends StatefulWidget {
...
+  /// highlight the search result.
+  ///
+  /// {@tool snippet}
+  ///
+  /// The following code shows a searchCallback to get an exact-match result:
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/dropdown_menu.dart:316]:
```suggestion
  /// In this example the `searchCallback` returns the index of the search result that exactly matches the query.
```

File: packages/flutter/lib/src/material/dropdown_menu.dart [hunk 3]
```diff
@@ -303,6 +308,34 @@ class DropdownMenu<T> extends StatefulWidget {
...
+  ///
+  ///     return index != -1 ? index : null;
+  ///   },
+  ///   dropdownMenuEntries: const <DropdownMenuEntry<Text>>[],
+  /// )
+  /// ```
+  /// {@end-tool}
+  ///
+  /// Defaults to null. If this is null and [DropdownMenu.enableSearch] is true,
+  /// the default function will return the first matching result which contains
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/dropdown_menu.dart:335]:
```suggestion
  /// the default function will return the index of the first matching result which contains the contents of the text input field.
```

File: packages/flutter/lib/src/material/dropdown_menu.dart [hunk 4]
```diff
@@ -21,6 +21,10 @@ import 'text_field.dart';
 import 'theme.dart';
 import 'theme_data.dart';
 
+/// A callback to search which item should be highlighted.
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/dropdown_menu.dart:24]:
```suggestion
/// A callback function that returns the index of the item that matches the current contents of a text field. 

/// If a match doesn't exist then null must be returned.
```

File: packages/flutter/lib/src/material/dropdown_menu.dart [hunk 5]
```diff
@@ -21,6 +21,10 @@ import 'text_field.dart';
...
+///
+/// Used by [DropdownMenu.searchCallback].
+typedef SearchCallback<T> = int? Function(List<DropdownMenuEntry<T>> entries, TextEditingController textEditingController);
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/dropdown_menu.dart:27]: Is there a reason to pass along the TextEditingController here? would it be enough to just pass along the controller's text value?

File: packages/flutter/lib/src/material/dropdown_menu.dart (hunk 1 above)
Comment (QuncCccccc) [reply packages/flutter/lib/src/material/dropdown_menu.dart:311]: Updated. Thanks!

File: packages/flutter/lib/src/material/dropdown_menu.dart (hunk 5 above)
Comment (QuncCccccc) [reply packages/flutter/lib/src/material/dropdown_menu.dart]: Ah Changing it to `String` sounds better. Just updated! Thanks for all the suggestions:) !

**Review by HansMuller — APPROVED** [review]
LGTM

**Pull request merged**